File: electron/launchersService.js
// Multi-Platform Game Launcher Service
// Detects Epic Games, GOG Galaxy, and EA App installations and games

import { exec } from 'child_process';
import { promisify } from 'util';
import fs from 'fs/promises';
import path from 'path';

const execAsync = promisify(exec);

// ============================================
// EPIC GAMES LAUNCHER
// ============================================

const EPIC_PATHS = [
    'C:\\Program Files\\Epic Games',
    'C:\\Program Files (x86)\\Epic Games',
    'D:\\Epic Games',
    'E:\\Epic Games'
];

const EPIC_LAUNCHER_PATHS = [
    process.env.LOCALAPPDATA + '\\EpicGamesLauncher',
    'C:\\Program Files (x86)\\Epic Games\\Launcher',
    'C:\\Program Files\\Epic Games\\Launcher'
];

/**
 * Find Epic Games Launcher installation
 */
async function findEpicPath() {
    // Try registry first
    try {
        const { stdout } = await execAsync(
            'reg query "HKEY_LOCAL_MACHINE\\SOFTWARE\\WOW6432Node\\Epic Games\\EpicGamesLauncher" /v AppDataPath',
            { encoding: 'utf8' }
        );
        const match = stdout.match(/AppDataPath\s+REG_SZ\s+(.+)/);
        if (match && match[1]) {
            return match[1].trim();
        }
    } catch (e) {
        // Registry not found
    }

    // Try common paths
    for (const epicPath of EPIC_LAUNCHER_PATHS) {
        try {
            await fs.access(epicPath);
            return epicPath;
        } catch { }
    }

    return null;
}

/**
 * Get installed Epic Games from manifest files
 */
async function getEpicInstalledGames() {
    const games = [];

    // Epic stores manifests in ProgramData
    const manifestPath = 'C:\\ProgramData\\Epic\\EpicGamesLauncher\\Data\\Manifests';

    try {
        const files = await fs.readdir(manifestPath);
        const itemFiles = files.filter(f => f.endsWith('.item'));

        for (const file of itemFiles) {
            try {
                const content = await fs.readFile(path.join(manifestPath, file), 'utf8');
                const manifest = JSON.parse(content);

                if (manifest.DisplayName && manifest.InstallLocation) {
                    games.push({
                        id: manifest.AppName || manifest.CatalogItemId,
                        name: manifest.DisplayName,
                        installPath: manifest.InstallLocation,
                        launchCommand: manifest.LaunchCommand || '',
                        executable: manifest.LaunchExecutable || '',
                        namespace: manifest.CatalogNamespace || '',
                        version: manifest.AppVersionString || '',
                        sizeOnDisk: manifest.InstallSize || 0,
                        isInstalled: true,
                        platform: 'epic'
                    });
                }
            } catch (e) {
                // Skip invalid manifest
            }
        }
    } catch (error) {
        console.error('Error reading Epic manifests:', error);
    }

    return games;
}

/**
 * Launch an Epic Games game
 */
function launchEpicGame(appName) {
    const launchUrl = `com.epicgames.launcher://apps/${appName}?action=launch&silent=true`;

    return new Promise((resolve, reject) => {
        exec(`start "" "${launchUrl}"`, (error) => {
            if (error) reject(error);
            else resolve({ success: true });
        });
    });
}

/**
 * Open Epic Games Store page
 */
function openEpicStore(appName) {
    const storeUrl = `com.epicgames.launcher://store/product/${appName}`;

    return new Promise((resolve, reject) => {
        exec(`start "" "${storeUrl}"`, (error) => {
            if (error) reject(error);
            else resolve({ success: true });
        });
    });
}

/**
 * Get Epic Games info
 */
async function getEpicInfo() {
    const launcherPath = await findEpicPath();
    const installedGames = await getEpicInstalledGames();

    return {
        success: true,
        installed: launcherPath !== null,
        launcherPath,
        games: installedGames,
        gameCount: installedGames.length
    };
}

// ============================================
// GOG GALAXY
// ============================================

const GOG_PATHS = [
    'C:\\Program Files (x86)\\GOG Galaxy',
    'C:\\Program Files\\GOG Galaxy',
    'D:\\GOG Galaxy',
    'E:\\GOG Galaxy'
];

/**
 * Find GOG Galaxy installation
 */
async function findGogPath() {
    // Try registry
    try {
        const { stdout } = await execAsync(
            'reg query "HKEY_LOCAL_MACHINE\\SOFTWARE\\WOW6432Node\\GOG.com\\GalaxyClient\\paths" /v client',
            { encoding: 'utf8' }
        );
        const match = stdout.match(/client\s+REG_SZ\s+(.+)/);
        if (match && match[1]) {
            return match[1].trim();
        }
    } catch (e) { }

    // Try common paths
    for (const gogPath of GOG_PATHS) {
        try {
            await fs.access(path.join(gogPath, 'GalaxyClient.exe'));
            return gogPath;
        } catch { }
    }

    return null;
}

/**
 * Get installed GOG games from database
 */
async function getGogInstalledGames() {
    const games = [];

    // GOG stores game info in ProgramData
    const gogDataPath = 'C:\\ProgramData\\GOG.com\\Galaxy\\storage\\galaxy-2.0.db';
    const gogGamesPath = process.env.LOCALAPPDATA + '\\GOG.com\\Galaxy\\Configuration\\InfoProviders\\GOG.ini';

    // Alternative: Read from installed games folder
    const gogGameFolders = [
        'C:\\GOG Games',
        'D:\\GOG Games',
        'C:\\Program Files (x86)\\GOG Galaxy\\Games',
        'C:\\Program Files\\GOG Galaxy\\Games'
    ];

    // Try to read from registry for installed games
    try {
        const { stdout } = await execAsync(
            'reg query "HKEY_LOCAL_MACHINE\\SOFTWARE\\WOW6432Node\\GOG.com\\Games" /s',
            { encoding: 'utf8' }
        );

        // Parse registry output to find games
        const gameBlocks = stdout.split(/HKEY_LOCAL_MACHINE\\SOFTWARE\\WOW6432Node\\GOG\.com\\Games\\/);

        for (const block of gameBlocks) {
            if (!block.trim()) continue;

            const gameIdMatch = block.match(/^(\d+)/);
            const nameMatch = block.match(/GAMENAME\s+REG_SZ\s+(.+)/i);
            const pathMatch = block.match(/PATH\s+REG_SZ\s+(.+)/i) || block.match(/EXE\s+REG_SZ\s+(.+)/i);
            const exeMatch = block.match(/EXE\s+REG_SZ\s+(.+)/i);

            if (gameIdMatch && nameMatch) {
                games.push({
                    id: gameIdMatch[1],
                    name: nameMatch[1].trim(),
                    installPath: pathMatch ? path.dirname(pathMatch[1].trim()) : '',
                    executable: exeMatch ? exeMatch[1].trim() : '',
                    isInstalled: true,
                    platform: 'gog'
                });
            }
        }
    } catch (e) {
        // Registry not accessible
    }

    // Scan game folders as fallback
    if (games.length === 0) {
        for (const folder of gogGameFolders) {
            try {
                const dirs = await fs.readdir(folder);
                for (const dir of dirs) {
                    const gamePath = path.join(folder, dir);
                    const stat = await fs.stat(gamePath);
                    if (stat.isDirectory()) {
                        // Check for goggame-*.info files
                        try {
                            const files = await fs.readdir(gamePath);
                            const infoFile = files.find(f => f.startsWith('goggame-') && f.endsWith('.info'));
                            if (infoFile) {
                                const infoContent = await fs.readFile(path.join(gamePath, infoFile), 'utf8');
                                const info = JSON.parse(infoContent);
                                games.push({
                                    id: info.gameId || infoFile.replace('goggame-', '').replace('.info', ''),
                                    name: info.name || dir,
                                    installPath: gamePath,
                                    executable: info.playTasks?.[0]?.path || '',
                                    isInstalled: true,
                                    platform: 'gog'
                                });
                            }
                        } catch { }
                    }
                }
            } catch { }
        }
    }

    return games;
}

/**
 * Launch a GOG game
 */
function launchGogGame(gameId) {
    const launchUrl = `goggalaxy://runGame/${gameId}`;

    return new Promise((resolve, reject) => {
        exec(`start "" "${launchUrl}"`, (error) => {
            if (error) reject(error);
            else resolve({ success: true });
        });
    });
}

/**
 * Get GOG Galaxy info
 */
async function getGogInfo() {
    const launcherPath = await findGogPath();
    const installedGames = await getGogInstalledGames();

    return {
        success: true,
        installed: launcherPath !== null,
        launcherPath,
        games: installedGames,
        gameCount: installedGames.length
    };
}

// ============================================
// EA APP (formerly Origin)
// ============================================

const EA_PATHS = [
    'C:\\Program Files\\Electronic Arts',
    'C:\\Program Files (x86)\\Origin',
    'C:\\Program Files\\EA Desktop',
    'C:\\Program Files (x86)\\EA Desktop',
    process.env.LOCALAPPDATA + '\\Electronic Arts',
    process.env.LOCALAPPDATA + '\\Origin'
];

/**
 * Find EA App installation
 */
async function findEaPath() {
    // Try EA App first (newer)
    try {
        const { stdout } = await execAsync(
            'reg query "HKEY_LOCAL_MACHINE\\SOFTWARE\\Electronic Arts\\EA Desktop" /v InstallLocation',
            { encoding: 'utf8' }
        );
        const match = stdout.match(/InstallLocation\s+REG_SZ\s+(.+)/);
        if (match && match[1]) {
            return { path: match[1].trim(), launcher: 'ea' };
        }
    } catch (e) { }

    // Try Origin as fallback
    try {
        const { stdout } = await execAsync(
            'reg query "HKEY_LOCAL_MACHINE\\SOFTWARE\\WOW6432Node\\Origin" /v ClientPath',
            { encoding: 'utf8' }
        );
        const match = stdout.match(/ClientPath\s+REG_SZ\s+(.+)/);
        if (match && match[1]) {
            return { path: path.dirname(match[1].trim()), launcher: 'origin' };
        }
    } catch (e) { }

    // Try common paths
    for (const eaPath of EA_PATHS) {
        try {
            await fs.access(eaPath);
            return { path: eaPath, launcher: 'ea' };
        } catch { }
    }

    return null;
}

/**
 * Get installed EA/Origin games
 */
async function getEaInstalledGames() {
    const games = [];

    // EA App stores game info in LocalContent
    const eaContentPath = 'C:\\ProgramData\\EA Desktop\\LocalContent';
    const originContentPath = 'C:\\ProgramData\\Origin\\LocalContent';

    const contentPaths = [eaContentPath, originContentPath];

    for (const contentPath of contentPaths) {
        try {
            const dirs = await fs.readdir(contentPath);

            for (const dir of dirs) {
                const gamePath = path.join(contentPath, dir);
                const stat = await fs.stat(gamePath);

                if (stat.isDirectory()) {
                    // Look for __Installer manifest
                    try {
                        const installerPath = path.join(gamePath, '__Installer');
                        const files = await fs.readdir(installerPath);
                        const manifestFile = files.find(f => f.endsWith('.mfst'));

                        if (manifestFile) {
                            const manifestContent = await fs.readFile(
                                path.join(installerPath, manifestFile),
                                'utf8'
                            );

                            // Parse manifest (it's in a custom format)
                            const idMatch = manifestContent.match(/id=([^\r\n&]+)/);

                            games.push({
                                id: idMatch ? idMatch[1] : dir,
                                name: dir.replace(/_/g, ' '),
                                installPath: gamePath,
                                isInstalled: true,
                                platform: 'ea'
                            });
                        }
                    } catch { }
                }
            }
        } catch { }
    }

    // Also check for games in standard EA Games folder
    const eaGamesFolders = [
        'C:\\Program Files\\EA Games',
        'C:\\Program Files (x86)\\EA Games',
        'D:\\EA Games',
        'C:\\Program Files\\Electronic Arts',
        'C:\\Program Files (x86)\\Electronic Arts'
    ];

    for (const folder of eaGamesFolders) {
        try {
            const dirs = await fs.readdir(folder);
            for (const dir of dirs) {
                const gamePath = path.join(folder, dir);
                const stat = await fs.stat(gamePath);
                if (stat.isDirectory() && !games.find(g => g.name === dir)) {
                    games.push({
                        id: dir.toLowerCase().replace(/\s+/g, '-'),
                        name: dir,
                        installPath: gamePath,
                        isInstalled: true,
                        platform: 'ea'
                    });
                }
            }
        } catch { }
    }

    return games;
}

/**
 * Launch an EA/Origin game
 */
function launchEaGame(gameId) {
    // Try EA App first, fallback to Origin
    const launchUrl = `origin://launchgame/${gameId}`;

    return new Promise((resolve, reject) => {
        exec(`start "" "${launchUrl}"`, (error) => {
            if (error) reject(error);
            else resolve({ success: true });
        });
    });
}

/**
 * Get EA App info
 */
async function getEaInfo() {
    const launcherInfo = await findEaPath();
    const installedGames = await getEaInstalledGames();

    return {
        success: true,
        installed: launcherInfo !== null,
        launcherPath: launcherInfo?.path || null,
        launcherType: launcherInfo?.launcher || null,
        games: installedGames,
        gameCount: installedGames.length
    };
}

// ============================================
// UBISOFT CONNECT (Bonus)
// ============================================

const UBISOFT_PATHS = [
    'C:\\Program Files (x86)\\Ubisoft\\Ubisoft Game Launcher',
    'C:\\Program Files\\Ubisoft\\Ubisoft Game Launcher',
    process.env.LOCALAPPDATA + '\\Ubisoft Game Launcher'
];

/**
 * Find Ubisoft Connect installation
 */
async function findUbisoftPath() {
    try {
        const { stdout } = await execAsync(
            'reg query "HKEY_LOCAL_MACHINE\\SOFTWARE\\WOW6432Node\\Ubisoft\\Launcher" /v InstallDir',
            { encoding: 'utf8' }
        );
        const match = stdout.match(/InstallDir\s+REG_SZ\s+(.+)/);
        if (match && match[1]) {
            return match[1].trim();
        }
    } catch (e) { }

    for (const ubisoftPath of UBISOFT_PATHS) {
        try {
            await fs.access(ubisoftPath);
            return ubisoftPath;
        } catch { }
    }

    return null;
}

/**
 * Get installed Ubisoft games
 */
async function getUbisoftInstalledGames() {
    const games = [];

    // Ubisoft stores install info in registry
    try {
        const { stdout } = await execAsync(
            'reg query "HKEY_LOCAL_MACHINE\\SOFTWARE\\WOW6432Node\\Ubisoft\\Launcher\\Installs" /s',
            { encoding: 'utf8' }
        );

        const gameBlocks = stdout.split(/HKEY_LOCAL_MACHINE\\SOFTWARE\\WOW6432Node\\Ubisoft\\Launcher\\Installs\\/);

        for (const block of gameBlocks) {
            if (!block.trim()) continue;

            const gameIdMatch = block.match(/^(\d+)/);
            const pathMatch = block.match(/InstallDir\s+REG_SZ\s+(.+)/i);

            if (gameIdMatch && pathMatch) {
                const installPath = pathMatch[1].trim();
                const gameName = path.basename(installPath);

                games.push({
                    id: gameIdMatch[1],
                    name: gameName,
                    installPath: installPath,
                    isInstalled: true,
                    platform: 'ubisoft'
                });
            }
        }
    } catch (e) {
        // Registry not accessible
    }

    return games;
}

/**
 * Launch a Ubisoft game
 */
function launchUbisoftGame(gameId) {
    const launchUrl = `uplay://launch/${gameId}/0`;

    return new Promise((resolve, reject) => {
        exec(`start "" "${launchUrl}"`, (error) => {
            if (error) reject(error);
            else resolve({ success: true });
        });
    });
}

/**
 * Get Ubisoft Connect info
 */
async function getUbisoftInfo() {
    const launcherPath = await findUbisoftPath();
    const installedGames = await getUbisoftInstalledGames();

    return {
        success: true,
        installed: launcherPath !== null,
        launcherPath,
        games: installedGames,
        gameCount: installedGames.length
    };
}

// ============================================
// COMBINED FUNCTIONS
// ============================================

/**
 * Get all launcher info at once
 */
async function getAllLaunchersInfo() {
    const [epic, gog, ea, ubisoft] = await Promise.all([
        getEpicInfo(),
        getGogInfo(),
        getEaInfo(),
        getUbisoftInfo()
    ]);

    return {
        epic,
        gog,
        ea,
        ubisoft,
        totalGames: epic.gameCount + gog.gameCount + ea.gameCount + ubisoft.gameCount
    };
}

/**
 * Launch a game from any platform
 */
async function launchGame(platform, gameId) {
    switch (platform) {
        case 'epic':
            return launchEpicGame(gameId);
        case 'gog':
            return launchGogGame(gameId);
        case 'ea':
        case 'origin':
            return launchEaGame(gameId);
        case 'ubisoft':
            return launchUbisoftGame(gameId);
        default:
            throw new Error(`Unknown platform: ${platform}`);
    }
}

export const launchersService = {
    // Epic Games
    getEpicInfo,
    getEpicInstalledGames,
    launchEpicGame,
    openEpicStore,

    // GOG Galaxy
    getGogInfo,
    getGogInstalledGames,
    launchGogGame,

    // EA App
    getEaInfo,
    getEaInstalledGames,
    launchEaGame,

    // Ubisoft Connect
    getUbisoftInfo,
    getUbisoftInstalledGames,
    launchUbisoftGame,

    // Combined
    getAllLaunchersInfo,
    launchGame
};

File: Orbit.Core/Services/LaunchersService.cs
using System.Diagnostics;
using System.IO;
using System.Text.Json;
using System.Text.RegularExpressions;
using Microsoft.Win32;

namespace Orbit.Core.Services;

/// <summary>
/// Multi-Platform Game Launcher Service
/// Detects Epic Games, GOG Galaxy, EA App, and Ubisoft Connect installations
/// </summary>
public class LaunchersService
{
    #region Common Classes

    public class InstalledGame
    {
        public string? Id { get; set; }
        public string? Name { get; set; }
        public string? InstallPath { get; set; }
        public string? Executable { get; set; }
        public bool IsInstalled { get; set; } = true;
        public string? Platform { get; set; }
        public string? LaunchCommand { get; set; }
        public long SizeOnDisk { get; set; }
    }

    public class LauncherInfo
    {
        public bool Success { get; set; } = true;
        public bool Installed { get; set; }
        public string? LauncherPath { get; set; }
        public List<InstalledGame> Games { get; set; } = new();
        public int GameCount => Games.Count;
    }

    #endregion

    #region Epic Games

    private static readonly string[] EPIC_LAUNCHER_PATHS = {
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "EpicGamesLauncher"),
        @"C:\Program Files (x86)\Epic Games\Launcher",
        @"C:\Program Files\Epic Games\Launcher"
    };

    public async Task<string?> FindEpicPath()
    {
        // Try registry
        try
        {
            using var key = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\WOW6432Node\Epic Games\EpicGamesLauncher");
            if (key?.GetValue("AppDataPath") is string path)
                return path;
        }
        catch { }

        // Try common paths
        foreach (var epicPath in EPIC_LAUNCHER_PATHS)
        {
            if (Directory.Exists(epicPath))
                return epicPath;
        }

        return null;
    }

    public async Task<List<InstalledGame>> GetEpicInstalledGames()
    {
        var games = new List<InstalledGame>();
        var manifestPath = @"C:\ProgramData\Epic\EpicGamesLauncher\Data\Manifests";

        try
        {
            if (!Directory.Exists(manifestPath))
                return games;

            var files = Directory.GetFiles(manifestPath, "*.item");

            foreach (var file in files)
            {
                try
                {
                    var content = await File.ReadAllTextAsync(file);
                    using var doc = JsonDocument.Parse(content);
                    var root = doc.RootElement;

                    var displayName = root.TryGetProperty("DisplayName", out var dn) ? dn.GetString() : null;
                    var installLocation = root.TryGetProperty("InstallLocation", out var il) ? il.GetString() : null;

                    if (!string.IsNullOrEmpty(displayName) && !string.IsNullOrEmpty(installLocation))
                    {
                        games.Add(new InstalledGame
                        {
                            Id = root.TryGetProperty("AppName", out var an) ? an.GetString() 
                                : root.TryGetProperty("CatalogItemId", out var cid) ? cid.GetString() : null,
                            Name = displayName,
                            InstallPath = installLocation,
                            Executable = root.TryGetProperty("LaunchExecutable", out var le) ? le.GetString() : null,
                            LaunchCommand = root.TryGetProperty("LaunchCommand", out var lc) ? lc.GetString() : null,
                            SizeOnDisk = root.TryGetProperty("InstallSize", out var isize) ? isize.GetInt64() : 0,
                            Platform = "epic"
                        });
                    }
                }
                catch { /* Skip invalid manifest */ }
            }
        }
        catch { }

        return games;
    }

    public async Task<object> LaunchEpicGame(string appName)
    {
        try
        {
            var launchUrl = $"com.epicgames.launcher://apps/{appName}?action=launch&silent=true";
            Process.Start(new ProcessStartInfo
            {
                FileName = launchUrl,
                UseShellExecute = true
            });
            return new { success = true };
        }
        catch (Exception ex)
        {
            return new { success = false, error = ex.Message };
        }
    }

    public async Task<object> OpenEpicStore(string appName)
    {
        try
        {
            var storeUrl = $"com.epicgames.launcher://store/product/{appName}";
            Process.Start(new ProcessStartInfo
            {
                FileName = storeUrl,
                UseShellExecute = true
            });
            return new { success = true };
        }
        catch (Exception ex)
        {
            return new { success = false, error = ex.Message };
        }
    }

    public async Task<object> GetEpicInfo()
    {
        var launcherPath = await FindEpicPath();
        var games = await GetEpicInstalledGames();

        return new
        {
            success = true,
            installed = launcherPath != null,
            launcherPath,
            games = games.Select(g => new
            {
                id = g.Id,
                name = g.Name,
                installPath = g.InstallPath,
                executable = g.Executable,
                isInstalled = g.IsInstalled,
                platform = g.Platform,
                sizeOnDisk = g.SizeOnDisk
            }).ToList(),
            gameCount = games.Count
        };
    }

    #endregion

    #region GOG Galaxy

    private static readonly string[] GOG_PATHS = {
        @"C:\Program Files (x86)\GOG Galaxy",
        @"C:\Program Files\GOG Galaxy",
        @"D:\GOG Galaxy",
        @"E:\GOG Galaxy"
    };

    private static readonly string[] GOG_GAME_FOLDERS = {
        @"C:\GOG Games",
        @"D:\GOG Games",
        @"C:\Program Files (x86)\GOG Galaxy\Games",
        @"C:\Program Files\GOG Galaxy\Games"
    };

    public async Task<string?> FindGogPath()
    {
        // Try registry
        try
        {
            using var key = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\WOW6432Node\GOG.com\GalaxyClient\paths");
            if (key?.GetValue("client") is string path)
                return path;
        }
        catch { }

        // Try common paths
        foreach (var gogPath in GOG_PATHS)
        {
            if (File.Exists(Path.Combine(gogPath, "GalaxyClient.exe")))
                return gogPath;
        }

        return null;
    }

    public async Task<List<InstalledGame>> GetGogInstalledGames()
    {
        var games = new List<InstalledGame>();

        // Try registry
        try
        {
            using var key = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\WOW6432Node\GOG.com\Games");
            if (key != null)
            {
                foreach (var subKeyName in key.GetSubKeyNames())
                {
                    try
                    {
                        using var gameKey = key.OpenSubKey(subKeyName);
                        if (gameKey != null)
                        {
                            var gameName = gameKey.GetValue("GAMENAME") as string;
                            var exePath = gameKey.GetValue("EXE") as string;
                            var installPath = gameKey.GetValue("PATH") as string ?? 
                                (exePath != null ? Path.GetDirectoryName(exePath) : null);

                            if (!string.IsNullOrEmpty(gameName))
                            {
                                games.Add(new InstalledGame
                                {
                                    Id = subKeyName,
                                    Name = gameName,
                                    InstallPath = installPath,
                                    Executable = exePath,
                                    Platform = "gog"
                                });
                            }
                        }
                    }
                    catch { }
                }
            }
        }
        catch { }

        // Scan game folders as fallback
        if (games.Count == 0)
        {
            foreach (var folder in GOG_GAME_FOLDERS)
            {
                if (!Directory.Exists(folder)) continue;

                try
                {
                    foreach (var dir in Directory.GetDirectories(folder))
                    {
                        try
                        {
                            var infoFiles = Directory.GetFiles(dir, "goggame-*.info");
                            if (infoFiles.Length > 0)
                            {
                                var content = await File.ReadAllTextAsync(infoFiles[0]);
                                using var doc = JsonDocument.Parse(content);
                                var root = doc.RootElement;

                                var gameId = root.TryGetProperty("gameId", out var gid) ? gid.GetString() 
                                    : Path.GetFileNameWithoutExtension(infoFiles[0]).Replace("goggame-", "");
                                var name = root.TryGetProperty("name", out var n) ? n.GetString() : Path.GetFileName(dir);

                                games.Add(new InstalledGame
                                {
                                    Id = gameId,
                                    Name = name,
                                    InstallPath = dir,
                                    Platform = "gog"
                                });
                            }
                        }
                        catch { }
                    }
                }
                catch { }
            }
        }

        return games;
    }

    public async Task<object> LaunchGogGame(string gameId)
    {
        try
        {
            var launchUrl = $"goggalaxy://runGame/{gameId}";
            Process.Start(new ProcessStartInfo
            {
                FileName = launchUrl,
                UseShellExecute = true
            });
            return new { success = true };
        }
        catch (Exception ex)
        {
            return new { success = false, error = ex.Message };
        }
    }

    public async Task<object> GetGogInfo()
    {
        var launcherPath = await FindGogPath();
        var games = await GetGogInstalledGames();

        return new
        {
            success = true,
            installed = launcherPath != null,
            launcherPath,
            games = games.Select(g => new
            {
                id = g.Id,
                name = g.Name,
                installPath = g.InstallPath,
                executable = g.Executable,
                isInstalled = g.IsInstalled,
                platform = g.Platform
            }).ToList(),
            gameCount = games.Count
        };
    }

    #endregion

    #region EA App / Origin

    private static readonly string[] EA_PATHS = {
        @"C:\Program Files\Electronic Arts",
        @"C:\Program Files (x86)\Origin",
        @"C:\Program Files\EA Desktop",
        @"C:\Program Files (x86)\EA Desktop",
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Electronic Arts"),
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Origin")
    };

    private static readonly string[] EA_GAME_FOLDERS = {
        @"C:\Program Files\EA Games",
        @"C:\Program Files (x86)\EA Games",
        @"D:\EA Games",
        @"C:\Program Files\Electronic Arts",
        @"C:\Program Files (x86)\Electronic Arts"
    };

    public async Task<(string? path, string? launcher)?> FindEaPath()
    {
        // Try EA Desktop
        try
        {
            using var key = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\Electronic Arts\EA Desktop");
            if (key?.GetValue("InstallLocation") is string path)
                return (path, "ea");
        }
        catch { }

        // Try Origin
        try
        {
            using var key = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\WOW6432Node\Origin");
            if (key?.GetValue("ClientPath") is string clientPath)
                return (Path.GetDirectoryName(clientPath), "origin");
        }
        catch { }

        // Try common paths
        foreach (var eaPath in EA_PATHS)
        {
            if (Directory.Exists(eaPath))
                return (eaPath, "ea");
        }

        return null;
    }

    public async Task<List<InstalledGame>> GetEaInstalledGames()
    {
        var games = new List<InstalledGame>();
        var contentPaths = new[] {
            @"C:\ProgramData\EA Desktop\LocalContent",
            @"C:\ProgramData\Origin\LocalContent"
        };

        foreach (var contentPath in contentPaths)
        {
            if (!Directory.Exists(contentPath)) continue;

            try
            {
                foreach (var dir in Directory.GetDirectories(contentPath))
                {
                    try
                    {
                        var installerPath = Path.Combine(dir, "__Installer");
                        if (!Directory.Exists(installerPath)) continue;

                        var manifestFiles = Directory.GetFiles(installerPath, "*.mfst");
                        if (manifestFiles.Length > 0)
                        {
                            var content = await File.ReadAllTextAsync(manifestFiles[0]);
                            var idMatch = Regex.Match(content, @"id=([^\r\n&]+)");

                            games.Add(new InstalledGame
                            {
                                Id = idMatch.Success ? idMatch.Groups[1].Value : Path.GetFileName(dir),
                                Name = Path.GetFileName(dir).Replace("_", " "),
                                InstallPath = dir,
                                Platform = "ea"
                            });
                        }
                    }
                    catch { }
                }
            }
            catch { }
        }

        // Check EA Games folders
        foreach (var folder in EA_GAME_FOLDERS)
        {
            if (!Directory.Exists(folder)) continue;

            try
            {
                foreach (var dir in Directory.GetDirectories(folder))
                {
                    var dirName = Path.GetFileName(dir);
                    if (!games.Any(g => g.Name == dirName))
                    {
                        games.Add(new InstalledGame
                        {
                            Id = dirName.ToLower().Replace(" ", "-"),
                            Name = dirName,
                            InstallPath = dir,
                            Platform = "ea"
                        });
                    }
                }
            }
            catch { }
        }

        return games;
    }

    public async Task<object> LaunchEaGame(string gameId)
    {
        try
        {
            var launchUrl = $"origin://launchgame/{gameId}";
            Process.Start(new ProcessStartInfo
            {
                FileName = launchUrl,
                UseShellExecute = true
            });
            return new { success = true };
        }
        catch (Exception ex)
        {
            return new { success = false, error = ex.Message };
        }
    }

    public async Task<object> GetEaInfo()
    {
        var launcherInfo = await FindEaPath();
        var games = await GetEaInstalledGames();

        return new
        {
            success = true,
            installed = launcherInfo != null,
            launcherPath = launcherInfo?.path,
            launcherType = launcherInfo?.launcher,
            games = games.Select(g => new
            {
                id = g.Id,
                name = g.Name,
                installPath = g.InstallPath,
                isInstalled = g.IsInstalled,
                platform = g.Platform
            }).ToList(),
            gameCount = games.Count
        };
    }

    #endregion

    #region Ubisoft Connect

    private static readonly string[] UBISOFT_PATHS = {
        @"C:\Program Files (x86)\Ubisoft\Ubisoft Game Launcher",
        @"C:\Program Files\Ubisoft\Ubisoft Game Launcher",
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Ubisoft Game Launcher")
    };

    public async Task<string?> FindUbisoftPath()
    {
        try
        {
            using var key = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\WOW6432Node\Ubisoft\Launcher");
            if (key?.GetValue("InstallDir") is string path)
                return path;
        }
        catch { }

        foreach (var ubisoftPath in UBISOFT_PATHS)
        {
            if (Directory.Exists(ubisoftPath))
                return ubisoftPath;
        }

        return null;
    }

    public async Task<List<InstalledGame>> GetUbisoftInstalledGames()
    {
        var games = new List<InstalledGame>();

        try
        {
            using var key = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\WOW6432Node\Ubisoft\Launcher\Installs");
            if (key != null)
            {
                foreach (var subKeyName in key.GetSubKeyNames())
                {
                    try
                    {
                        using var gameKey = key.OpenSubKey(subKeyName);
                        if (gameKey?.GetValue("InstallDir") is string installPath)
                        {
                            games.Add(new InstalledGame
                            {
                                Id = subKeyName,
                                Name = Path.GetFileName(installPath),
                                InstallPath = installPath,
                                Platform = "ubisoft"
                            });
                        }
                    }
                    catch { }
                }
            }
        }
        catch { }

        return games;
    }

    public async Task<object> LaunchUbisoftGame(string gameId)
    {
        try
        {
            var launchUrl = $"uplay://launch/{gameId}/0";
            Process.Start(new ProcessStartInfo
            {
                FileName = launchUrl,
                UseShellExecute = true
            });
            return new { success = true };
        }
        catch (Exception ex)
        {
            return new { success = false, error = ex.Message };
        }
    }

    public async Task<object> GetUbisoftInfo()
    {
        var launcherPath = await FindUbisoftPath();
        var games = await GetUbisoftInstalledGames();

        return new
        {
            success = true,
            installed = launcherPath != null,
            launcherPath,
            games = games.Select(g => new
            {
                id = g.Id,
                name = g.Name,
                installPath = g.InstallPath,
                isInstalled = g.IsInstalled,
                platform = g.Platform
            }).ToList(),
            gameCount = games.Count
        };
    }

    #endregion

    #region Combined Functions

    public async Task<object> GetAllLaunchersInfo()
    {
        var epicTask = GetEpicInfo();
        var gogTask = GetGogInfo();
        var eaTask = GetEaInfo();
        var ubisoftTask = GetUbisoftInfo();

        await Task.WhenAll(epicTask, gogTask, eaTask, ubisoftTask);

        var epic = await epicTask;
        var gog = await gogTask;
        var ea = await eaTask;
        var ubisoft = await ubisoftTask;

        // Extract game counts using reflection or dynamic
        var epicJson = JsonSerializer.Serialize(epic);
        var gogJson = JsonSerializer.Serialize(gog);
        var eaJson = JsonSerializer.Serialize(ea);
        var ubisoftJson = JsonSerializer.Serialize(ubisoft);

        using var epicDoc = JsonDocument.Parse(epicJson);
        using var gogDoc = JsonDocument.Parse(gogJson);
        using var eaDoc = JsonDocument.Parse(eaJson);
        using var ubisoftDoc = JsonDocument.Parse(ubisoftJson);

        var totalGames = 
            (epicDoc.RootElement.TryGetProperty("gameCount", out var ec) ? ec.GetInt32() : 0) +
            (gogDoc.RootElement.TryGetProperty("gameCount", out var gc) ? gc.GetInt32() : 0) +
            (eaDoc.RootElement.TryGetProperty("gameCount", out var eac) ? eac.GetInt32() : 0) +
            (ubisoftDoc.RootElement.TryGetProperty("gameCount", out var uc) ? uc.GetInt32() : 0);

        return new
        {
            epic,
            gog,
            ea,
            ubisoft,
            totalGames
        };
    }

    public async Task<object> LaunchGame(string platform, string gameId)
    {
        return platform.ToLower() switch
        {
            "epic" => await LaunchEpicGame(gameId),
            "gog" => await LaunchGogGame(gameId),
            "ea" or "origin" => await LaunchEaGame(gameId),
            "ubisoft" => await LaunchUbisoftGame(gameId),
            _ => new { success = false, error = $"Unknown platform: {platform}" }
        };
    }

    #endregion
}
